decision_service/dsclientlib/error_callback_fn.cc

```diff
@@ -0,0 +1,36 @@
+#include "error_callback_fn.h"
+
+using namespace std;
+namespace decision_service {
+
+  void error_callback_fn::set(error_fn fn, void* context)
+  {
+    lock_guard<mutex> lock(_mutex);
+    _fn = fn;
+    _context = context;
+  }
+
+  void error_callback_fn::report_error(api_status& s)
+  {
+    if (_fn == nullptr) 
+      return;
+
+    lock_guard<mutex> lock(_mutex);
```

Comment:
>lock_guard<mutex> lock(_mutex); [](start = 4, length = 31)

that looks heavy.

Reply:
a couple of issues here.  1) background thread is using the function pointer 2) we might have more than one background thread and user supplied error callback might not be reentrant. 

---
In reply to: [186737190](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186737190) [](ancestors = 186737190)